Code improvements

diveshuttam opened the pull request:
This closes #83, closes #65, closes #79 and closes #57 .

diveshuttam committed:
Improve code

closes #83 #79 #65 #57

diveshuttam commented:
This **needs** to be based on #81. I am closing this for now, will open again after #81 is completed.

dashohoxha commented on src/gpg/contact/delete.py:
```diff
@@ -21,10 +21,11 @@ def delete(contacts, force):
             raise
-        exit(2)
+        print(e, file=sys.stderr, flush=True)
+        exit(1)
```
These 4 lines seem to be the same everywhere. Why don't you create a utility (auxiliary) function to encapsulate them, for example: `handle_exception(e)`

sourabhtk37 replied:
Also shouldn't the print statement be in the else statement?

diveshuttam replied:
Umm I guess raise causes an exit as well.

dashohoxha commented on src/gpg/contact/export.py:
```diff
@@ -6,11 +6,16 @@
+            print("Not enough permissions to write to {path}".format(
+                path=export_path
+            ), file=sys.stderr, flush=True)
+            exit(1)
```
These lines seem to be the same everywhere. Why don't you simplify them with a utility function like this: 
```
print_error("Not enough permissions to write to {path}".format(path=export_path))
```

diveshuttam committed:
Merge branch 'gnupg-2.2' into code_improvements

sourabhtk37 commented on src/gpg/fn/interact.py:
```diff
@@ -14,18 +14,18 @@ def edit_fnc(self, status, args, out=None):
-                self.done = True
-        else:
-            cmd = None
+            self.done = len(self.cmds) == self.step
```
This doesn't seem that readable. Maybe try ternary operation for one-liner:
```python 
True if <expr> else False

sourabhtk37 commented on src/gpg/fn/print_key.py:
```diff
@@ -84,7 +89,7 @@ def print_key(identity, end=""):
     sign_list = set({})
```
`set()`

sourabhtk37 commented on src/gpg/key/gen.py:
```diff
@@ -4,23 +4,23 @@
+        params = """<GnupgKeyParms format="internal">
+        {parameters}
+        </GnupgKeyParms>
+        """.format(parameters=parameters)
```
Try block should start from line 13 imo i.e Piece of code that are likely to cause exceptions.

diveshuttam replied:
Yup but it doesn't change the behavior if we have correct lines in `try`. Also am am catching specific exceptions now so it won't mask any unknown exceptions.

sourabhtk37 commented on src/gpg/key/gen.py:
```diff
@@ -4,23 +4,23 @@
 
 
 def gen(parameters):
```
A bit more descriptive name helps in readability. `generate_key` or `gen_key`.

sourabhtk37 reviewed (commented):
Great work @diveshuttam

diveshuttam committed:
Use decorators and other functions to simplify code

diveshuttam commented:
I researched a bit and found about [decorators](https://www.python.org/dev/peps/pep-0318/). I have created handle_exception as a decorator. It wraps both try, except as well as raising/printing error functionality. It shortens code greatly and also looks prettier IMO. @sourabhtk37, @dashohoxha need your suggestions.

dashohoxha commented:
The definition of the decorator looks like magic to me. The rest looks great.

By the way, you have missed an 'i' in the spelling of 'auxiliary'.

diveshuttam committed:
s/auxilary/auxiliary

diveshuttam committed:
Add note about handle exception decorator

diveshuttam commented:
>The definition of the decorator looks like magic to me. 

I have added a comment about the decorator in the code.
Someone familiar with python must not have much problem. But still it helps if one understands code.

BTW just for reference:

```python
@decorator
def function():
    something
```

is equivalent to 
```python
def function():
    something

function = decorator(function)
```

I have used two levels of decoration in my case as I am passing exceptions.

> By the way, you have missed an 'i' in the spelling of 'auxiliary'.

Corrected this.

dashohoxha reviewed (approved):
It looks OK to me.

diveshuttam commented:
@sourabhtk37 please go through this once again before merging. Since this introduces a major change.

dashohoxha commented:
> @sourabhtk37 please go through this once again before merging. Since this introduces a major change.

@diveshuttam it is important that you run the tests and make sure that the changes don't break them.
The things that me and Sourabh can ask for changing are basically small things and not so important. I mean it is not so important whether you write **if** with parentheses or without parentheses, but it is very important if there is a bug.

diveshuttam commented:
Yes I ran the tests and they are passing. Just asking if the my methodology is right and maybe if @sourabhtk37 has something better in mind.

sourabhtk37 commented:
@diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

dashohoxha commented:
> @diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

It is not obvious to me what problems it may cause. So, let's keep as it is for the time being, and we can always fix it later in case we notice some problems.

diveshuttam commented:
>@diveshuttam The decorators approach is good but putting the entire function code inside a try-except block is not. It might cause unexpected behaviour being python. Just my 2 cents.

I have merged this for now. Though the entire function has `try except`, it will only exit as soon as first exception arises, so I think it shouldn't cause much problem. In case I wanted to do something else, and not exit, we can use inner `try except` also as the error binds to the closest `try except` first. I have 
 done similar thing in couple of places.
